service/worker/batcher/batcher.go

```diff
@@ -0,0 +1,174 @@
+func New(params *BootstrapParams) *Batcher {
+	cfg := params.Config
+	return &Batcher{
+		context: batcherContext{
+			cfg:           cfg,
+			svcClient:     params.ServiceClient,
+			metricsClient: params.MetricsClient,
+			tallyScope:    params.TallyScope,
+			logger:        params.Logger.WithTags(tag.ComponentBatcher),
+		},
+	}
+}
+
+// Start starts the scanner
+func (s *Batcher) Start() error {
+	//retry until making sure global system domain is there
+	err := s.createGlobalSystemDomainIfNotExistsWithRetry()
+	if err != nil {
+		return err
+	}
+
+	// start worker for batch operation workflows
+	workerOpts := worker.Options{
+		MetricsScope:              s.context.tallyScope,
+		BackgroundActivityContext: context.WithValue(context.Background(), batcherContextKey, s.context),
+	}
+	worker := worker.New(s.context.svcClient, common.SystemDomainName, batcherTaskListName, workerOpts)
+	return worker.Start()
+}
+
+func (s *Batcher) createGlobalSystemDomainIfNotExistsWithRetry() error {
+	policy := backoff.NewExponentialRetryPolicy(backOffInitialInterval)
+	policy.SetMaximumInterval(backOffMaxInterval)
+	policy.SetExpirationInterval(maxStartupTime)
+	return backoff.Retry(func() error {
+		return s.createGlobalSystemDomainIfNotExists()
+	}, policy, func(err error) bool {
+		return true
+	})
+}
+
+func (s *Batcher) createGlobalSystemDomainIfNotExists() error {
+	ctx, cancel := context.WithTimeout(context.Background(), rpcTimeout)
+	_, err := s.context.svcClient.DescribeDomain(ctx, &shared.DescribeDomainRequest{
+		Name: common.StringPtr(common.SystemGlobalDomainName),
+	})
+	cancel()
+
+	if err == nil {
+		s.context.logger.Info("Global system domain already exists", tag.WorkflowDomainName(common.SystemGlobalDomainName))
+		return nil
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), rpcTimeout)
+	err = s.context.svcClient.RegisterDomain(ctx, s.getDomainCreationRequest())
+	cancel()
+	if err != nil {
+		s.context.logger.Error("Error creating global system domain", tag.Error(err))
+		return err
+	}
+	s.context.logger.Info("Domain created successfully")
+	return nil
+}
+
+func (s *Batcher) getDomainCreationRequest() *shared.RegisterDomainRequest {
+	req := &shared.RegisterDomainRequest{
+		Name:                                   common.StringPtr(common.SystemGlobalDomainName),
+		WorkflowExecutionRetentionPeriodInDays: common.Int32Ptr(common.SystemDomainRetentionDays),
+		EmitMetric:                             common.BoolPtr(true),
+		SecurityToken:                          common.StringPtr(s.context.cfg.AdminOperationToken()),
+		IsGlobalDomain:                         common.BoolPtr(false),
+	}
+
+	if s.context.cfg.ClusterMetadata.IsGlobalDomainEnabled() {
+		req.IsGlobalDomain = common.BoolPtr(true)
+		req.ActiveClusterName = common.StringPtr(s.context.cfg.ClusterMetadata.GetMasterClusterName())
+		var clusters []*shared.ClusterReplicationConfiguration
+		for name := range s.context.cfg.ClusterMetadata.GetAllClusterInfo() {
```

Comment:
plz also check inside the cluster info whether the cluster is enabled

Reply:
Sure